File: core/src/main/java/nl/sjtek/control/core/network/ResponseCache.java
package nl.sjtek.control.core.network;

import com.google.common.eventbus.Subscribe;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import nl.sjtek.control.core.events.BroadcastEvent;
import nl.sjtek.control.core.events.Bus;
import nl.sjtek.control.core.events.DataChangedEvent;
import nl.sjtek.control.data.responses.Response;
import nl.sjtek.control.data.responses.ResponseAdapter;

import java.util.HashMap;
import java.util.Map;


public class ResponseCache {

    private static ResponseCache instance = new ResponseCache();
    private Map<String, Response> responseMap = new HashMap<>();

    private ResponseCache() {
        Bus.regsiter(this);
    }

    public synchronized static ResponseCache getInstance() {
        return instance;
    }

    public String toJson() {
        Gson gson = new GsonBuilder()
                .registerTypeHierarchyAdapter(Response.class, new ResponseAdapter())
                .setPrettyPrinting()
                .create();
        return gson.toJson(responseMap);
    }

    @Subscribe
    public void onUpdate(DataChangedEvent event) {
        this.responseMap.put(event.getKey(), event.getResponse());
        if (event.shouldPushToClients()) {
            Bus.post(new BroadcastEvent(toJson()));
        }
    }
}

File: core/src/main/kotlin/nl/sjtek/control/core/response/ResponseCache.kt
package nl.sjtek.control.core.response

import com.google.gson.GsonBuilder
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.FlowableEmitter
import nl.sjtek.control.core.events.BroadcastEvent
import nl.sjtek.control.core.events.Bus
import nl.sjtek.control.core.modules.Module
import nl.sjtek.control.data.response.Response
import org.slf4j.LoggerFactory
import java.util.concurrent.TimeUnit


object ResponseCache {

    private val logger = LoggerFactory.getLogger(javaClass)
    private val gson = GsonBuilder().setPrettyPrinting().create()
    private val responses: MutableMap<String, Response> = mutableMapOf()
    var json: String = "{}"
        private set
    private var previousJson: String = "{}"
    private var emitter: FlowableEmitter<Module>? = null

    init {
        Flowable.create({ e: FlowableEmitter<Module> ->
            emitter = e
        }, BackpressureStrategy.LATEST)
                .debounce(750, TimeUnit.MILLISECONDS)
                .subscribe { module ->
                    if (json != previousJson) {
                        previousJson = json
                        logger.info("Broadcasting update (${module.key})")
                        Bus.post(BroadcastEvent(json))
                    }
                }
    }

    fun getKeys(keys: List<String>): String {
        val result = responses.filter { it.key in keys }
        return gson.toJson(result)
    }

    @Synchronized
    fun post(module: Module, broadcast: Boolean = false) {
        responses.put(module.key, module.response)
        json = gson.toJson(responses)

        if (broadcast) emitter?.onNext(module)
    }
}